module/apmprometheus/gatherer.go

```diff
@@ -80,8 +80,70 @@ func (g gatherer) GatherMetrics(ctx context.Context, out *apm.Metrics) error {
 					out.Add(name+".percentile."+strconv.Itoa(p), labels, q.GetValue())
 				}
 			}
+		case dto.MetricType_HISTOGRAM:
+			// For the bucket values, we follow the approach described by Prometheus's
+			// histogram_quantile function (https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile)
+			// to achieve consistent percentile aggregation results:
+			//
+			// "The histogram_quantile() function interpolates quantile values by assuming a linear
+			// distribution within a bucket. (...) If a quantile is located in the highest bucket,
+			// the upper bound of the second highest bucket is returned. A lower limit of the lowest
+			// bucket is assumed to be 0 if the upper bound of that bucket is greater than 0. In that
+			// case, the usual linear interpolation is applied within that bucket. Otherwise, the upper
+			// bound of the lowest bucket is returned for quantiles located in the lowest bucket."
+			for _, m := range mf.GetMetric() {
+				h := m.GetHistogram()
+				// Total count for all values in this
+				// histogram. We want the per value count.
+				if h.GetSampleCount() == 0 {
+					continue
+				}
+				labels := makeLabels(m.GetLabel())
+				values := h.GetBucket()
+				midpoints := make([]float64, 0, len(values))
+				counts := make([]uint64, 0, len(values))
+				for i, b := range values {
+					count := b.GetCumulativeCount()
+					// we are excluding zero-count
+					// prometheus buckets.
+					if count == 0 {
+						continue
+					}
+					le := b.GetUpperBound()
+					if i == 0 {
+						if le > 0 {
+							le /= 2
+						}
+					} else if i == (len(values) - 1) {
+						le = values[i-1].GetUpperBound()
+						// apm-server expects non-cumulative
+						// counts. prometheus counts each
+						// bucket cumulatively, ie. bucketN
+						// contains all counts for bucketN and
+						// all counts in preceding values. To
+						// get the current bucket's count we
+						// subtract bucketN-1 from bucketN,
+						// when N>0.
+						count = count - values[i-1].GetCumulativeCount()
+					} else {
+						le = values[i-1].GetUpperBound() + (le-values[i-1].GetUpperBound())/2.0
+						// prometheus counts buckets cumulatively.
+						count = count - values[i-1].GetCumulativeCount()
+					}
+					// re-check count. the cumulative count
+					// in the initial check may have been
+					// non-zero, but when we subtract the
+					// preceding bucket, it may end up
+					// having a zero count.
```

Comment:
I wonder if there's enough value in the initial count check, wouldn't the code between be cheap? Maybe just  do the check once here?

Reply:
that's fine by me. also, the first check is of limited use since it only works until we get a bucket with a count, after which point all buckets' cumulative counts will be initially non-zero.